modules/openapi-generator/src/main/resources/typescript-fetch/apis.mustache

```diff
@@ -338,24 +338,24 @@ export class {{classname}} extends runtime.BaseAPI {
      {{/summary}}
      */
     {{^useSingleRequestParameter}}
-    async {{nickname}}({{#allParams}}{{paramName}}{{^required}}?{{/required}}: {{#isEnum}}{{{datatypeWithEnum}}}{{/isEnum}}{{^isEnum}}{{{dataType}}}{{#isNullable}} | null{{/isNullable}}{{/isEnum}}, {{/allParams}}initOverrides?: RequestInit | runtime.InitOverrideFunction): Promise<{{{returnType}}}{{^returnType}}void{{/returnType}}> {
+    async {{nickname}}({{#allParams}}{{paramName}}{{^required}}?{{/required}}: {{#isEnum}}{{{datatypeWithEnum}}}{{/isEnum}}{{^isEnum}}{{{dataType}}}{{#isNullable}} | null{{/isNullable}}{{/isEnum}}, {{/allParams}}initOverrides?: RequestInit | runtime.InitOverrideFunction, metadata?: any): Promise<{{{returnType}}}{{^returnType}}void{{/returnType}}> {
```

Comment:
can we find an alternative to an unnamed parameter with type `any`?

Reply:
The type of this configuration is driven by the user code base, which we do not know during the generation. The only way to provide meaningful type here I see is to generify `runtime.BaseAPI`, its implementations and `Middleware`. Then users can create instances of the APIs parametrized with their interceptor configuration types. We could make the default value for the generic type `void`, so the change is backwards compatible and existing clients do not need to change, e.g.:
```
export class BaseAPI<M = void> {
...
  protected async request(context: RequestOpts, initOverrides?: RequestInit | InitOverrideFunction, metadata?: M): Promise<Response> {
```

As for the parameter, my initial intention was to include it as a field into `initOverrides` object, but I was not sure if semantically it is correct. If this is a good option, then we need to change the parameter type from current `initOverrides?: RequestInit | InitOverrideFunction` to new `initOverrides?: ExtendedRequestInit | InitOverrideFunction`, where `ExtendedRequestInit` is something like this:
```
interface ExtendedRequestInit extends RequestInit {
  metadata?: any;
}
...
export type InitOverrideFunction = (requestContext: { init: HTTPRequestInit, context: RequestOpts }) => Promise<ExtendedRequestInit>
```
(and if we go with generics for `BaseAPI`, then `ExtendedRequestInit` should also be parametrized).
As far as I can see, this change would also be backwards compatible in TypeScript.

Wdyt, should I go with these two changes or do you have better solutions?

Reply:
Alright, so I decided to try the generics approach. It looks reasonably good to me, at least on the client side instead of untyped
```
await myApi.getSomething({}, { myConfig: myValue });
```
we can write more concise and strongly typed
```
await myApi.getSomething({ myConfig: myValue });
```
and interceptors are also strongly typed. 

I see two drawbacks in this option:
* The generated code is more verbose due to generic parameters. But this probably is not a big issue for the generated code.
* Formally backwards compatibility is broken for `InitOverrideFunction` - it now requires a type parameter (I could not find a way to use defaults). But this only affects users if they explicitly created a variable with defined type, e.g.:
  ```
  const myOverrides : InitOverrideFunction = async () => ....
  ...
  await myApi.getSomething(myOverrrides);
  ``` 
  Personally I would not use this notation, but cannot say for all users.

Waiting for the feedback.

Reply:
thanks a lot for the detailed investigation, and sorry for the late reply!
given that the each middleware might want to accept a different type of `metadata`, and we don't know anything about it at generation time, we might want to keep it as simple as possible, while still expressive, by e.g. changing the api function signature so something like
```typescript
myMethod(/* parameters, etc. ... */, additionalParameters: {metadata: any})
```
where we could move `{metadata: any}` to a separate interface, with some explanation what `metadata` is and how it is used.
This way, we avoid accepting a `myMethod(metadata: any)` parameter, and gain some safety on the caller side, which needs to explicitly provide the name when calling `myMethod({metadata: ...})`.

do you agree?

Reply:
Each interceptor accepting different type of metadata is definitely one of the options.

At the same time, from my (generator user) point of view, all interceptors are part of my application infrastructure and they are transparent to my business code. E.g. when making a request from the business code, I do not operate configuration of InterceptorA or InterceptorB, I configure the overall request behaviour, as per my application API client capabilities. This decouples business code from implementation details of the API client. 
With this in mind, my expectation is that request metadata is defined per application, not per interceptor. This approach also opens a possibility to make the metadata typed, making business code and infrastructure code type-safe (I am using this version in my project and from application side it looks nice and clean). 
What do you think, does it make sense?

Reply:
I've separated `initOverrides` and `additionalParameters`, this simplified the generator code to some extent.

Reply:
Hi guys, may I ask you to make a decision regarding this PR? Let's finalize and merge or close it.